### ISSUE TITLE
[Bug] Implied eval error not fixed

### ISSUE THREAD

**Issue by Kristinita** [eslint/eslint#7861]
### 1. Summary

My linter show error in my code: [**Implied eval. Consider passing a function instead of a string**](https://github.com/jamesallardice/jslint-error-explanations/blob/master/message-articles/implied-eval.md). ESLint have [**no-implied-eval**](http://eslint.org/docs/rules/no-implied-eval) rule, but not fixed it for me, I still get same error.

### 2. Expected behavior

Successfully fixed ``Implied eval error``.

### 3. Actual behavior

Not fixed ``Implied eval error``.

### 4. Settings

My ```.eslintrc.js``` file in ```E:\Киролайна``` folder:

<details>
<summary>
<b>Click here to see full file.</b></summary><br>

```js
module.exports = {
    "env": {
        "browser": true,
        "es6": true
    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "sourceType": "module"
    },
    "rules": {
        "accessor-pairs": "error",
        "array-bracket-spacing": "error",
        "array-callback-return": "error",
        "arrow-body-style": "error",
        "arrow-parens": "error",
        "arrow-spacing": "error",
        "block-scoped-var": "error",
        "block-spacing": "error",
        "brace-style": "error",
        "callback-return": "error",
        "camelcase": "error",
        "capitalized-comments": "error",
        "class-methods-use-this": "error",
        "comma-dangle": "error",
        "comma-spacing": [
            "error",
            {
                "after": true,
                "before": false
            }
        ],
        "comma-style": "error",
        "complexity": "error",
        "computed-property-spacing": [
            "error",
            "never"
        ],
        "consistent-return": "error",
        "consistent-this": "error",
        "curly": "error",
        "default-case": "error",
        "dot-location": "error",
        "dot-notation": "error",
        "eol-last": [
            "error",
            "never"
        ],
        "eqeqeq": "error",
        "func-call-spacing": "error",
        "func-name-matching": "error",
        "func-names": [
            "error",
            "never"
        ],
        "func-style": "error",
        "generator-star-spacing": "error",
        "global-require": "error",
        "guard-for-in": "error",
        "handle-callback-err": "error",
        "id-blacklist": "error",
        "id-length": "off",
        "id-match": "error",
        "indent": "off",
        "init-declarations": "error",
        "jsx-quotes": "error",
        "key-spacing": "error",
        "keyword-spacing": "error",
        "line-comment-position": "error",
        "linebreak-style": [
            "error",
            "unix"
        ],
        "lines-around-comment": "error",
        "lines-around-directive": "error",
        "max-depth": "error",
        "max-len": "error",
        "max-lines": "error",
        "max-nested-callbacks": "error",
        "max-params": "error",
        "max-statements": "error",
        "max-statements-per-line": "error",
        "multiline-ternary": "error",
        "new-cap": "error",
        "new-parens": "error",
        "newline-after-var": [
            "error",
            "never"
        ],
        "newline-before-return": "error",
        "newline-per-chained-call": "error",
        "no-alert": "error",
        "no-array-constructor": "error",
        "no-await-in-loop": "error",
        "no-bitwise": "error",
        "no-caller": "error",
        "no-catch-shadow": "error",
        "no-confusing-arrow": "error",
        "no-continue": "error",
        "no-div-regex": "error",
        "no-duplicate-imports": "error",
        "no-else-return": "error",
        "no-empty-function": "error",
        "no-eq-null": "error",
        "no-eval": "error",
        "no-extend-native": "error",
        "no-extra-bind": "error",
        "no-extra-label": "error",
        "no-extra-parens": "error",
        "no-floating-decimal": "error",
        "no-implicit-coercion": "error",
        "no-implicit-globals": "error",
        "no-implied-eval": "off",
        "no-inline-comments": "error",
        "no-invalid-this": "error",
        "no-iterator": "error",
        "no-label-var": "error",
        "no-labels": "error",
        "no-lone-blocks": "error",
        "no-lonely-if": "error",
        "no-loop-func": "error",
        "no-magic-numbers": "off",
        "no-mixed-operators": "error",
        "no-mixed-requires": "error",
        "no-multi-spaces": "error",
        "no-multi-str": "error",
        "no-multiple-empty-lines": "error",
        "no-native-reassign": "error",
        "no-negated-condition": "error",
        "no-negated-in-lhs": "error",
        "no-nested-ternary": "error",
        "no-new": "error",
        "no-new-func": "error",
        "no-new-object": "error",
        "no-new-require": "error",
        "no-new-wrappers": "error",
        "no-octal-escape": "error",
        "no-param-reassign": "error",
        "no-path-concat": "error",
        "no-plusplus": "error",
        "no-process-env": "error",
        "no-process-exit": "error",
        "no-proto": "error",
        "no-prototype-builtins": "error",
        "no-restricted-globals": "error",
        "no-restricted-imports": "error",
        "no-restricted-modules": "error",
        "no-restricted-properties": "error",
        "no-restricted-syntax": "error",
        "no-return-assign": "error",
        "no-return-await": "error",
        "no-script-url": "error",
        "no-self-compare": "error",
        "no-sequences": "error",
        "no-shadow": "error",
        "no-shadow-restricted-names": "error",
        "no-spaced-func": "error",
        "no-sync": "error",
        "no-tabs": "error",
        "no-template-curly-in-string": "error",
        "no-ternary": "error",
        "no-throw-literal": "error",
        "no-trailing-spaces": "error",
        "no-undef-init": "error",
        "no-undefined": "error",
        "no-underscore-dangle": "error",
        "no-unmodified-loop-condition": "error",
        "no-unneeded-ternary": "error",
        "no-unused-expressions": "error",
        "no-use-before-define": "error",
        "no-useless-call": "error",
        "no-useless-computed-key": "error",
        "no-useless-concat": "error",
        "no-useless-constructor": "error",
        "no-useless-escape": "error",
        "no-useless-rename": "error",
        "no-useless-return": "error",
        "no-var": "off",
        "no-void": "error",
        "no-warning-comments": "error",
        "no-whitespace-before-property": "error",
        "no-with": "error",
        "object-curly-newline": "error",
        "object-curly-spacing": "error",
        "object-property-newline": "error",
        "object-shorthand": "error",
        "one-var": "off",
        "one-var-declaration-per-line": "error",
        "operator-assignment": "error",
        "operator-linebreak": "error",
        "padded-blocks": "off",
        "prefer-arrow-callback": "error",
        "prefer-const": "error",
        "prefer-numeric-literals": "error",
        "prefer-reflect": "error",
        "prefer-rest-params": "error",
        "prefer-spread": "error",
        "prefer-template": "off",
        "quote-props": "error",
        "quotes": [
            "error",
            "single"
        ],
        "radix": "error",
        "require-await": "error",
        "require-jsdoc": "error",
        "rest-spread-spacing": "error",
        "semi": "error",
        "semi-spacing": "error",
        "sort-imports": "error",
        "sort-keys": "error",
        "sort-vars": "error",
        "space-before-blocks": "error",
        "space-before-function-paren": "error",
        "space-in-parens": [
            "error",
            "never"
        ],
        "space-infix-ops": "error",
        "space-unary-ops": "error",
        "spaced-comment": "error",
        "strict": "error",
        "symbol-description": "error",
        "template-curly-spacing": "error",
        "unicode-bom": [
            "error",
            "never"
        ],
        "valid-jsdoc": "error",
        "vars-on-top": "off",
        "wrap-iife": "error",
        "wrap-regex": "error",
        "yield-star-spacing": "error",
        "yoda": "error"
    }
};
```
</details><br>

### 5. Steps to reproduce

I have [**Google Custom Search Engine**](https://cse.google.com) search system. [**autofocus**](http://www.w3schools.com/tags/att_input_autofocus.asp) attribute not working for GCSE input field. I find [**solution**](https://www.experts-exchange.com/questions/27924508/How-to-get-autofocus-in-the-search-field-using-Google's-Search-Engine-on-a-site.html#a38565775) in Expert Exchange site — I add line

```js
setTimeout("document.getElementById('gsc-i-id1').focus();", 1000);
```

in my script, but I get error [**Implied eval. Consider passing a function instead of a string**](https://github.com/jamesallardice/jslint-error-explanations/blob/master/message-articles/implied-eval.md). I can not fix error manually and I try to use ESLinter.

`npm install eslint -g` → `eslint init` in `Киролайна` folder → I set my `.eslintrc.js` file.

My `test.js` file before fixing:

```js
(function() {
    var cx = '013024336414733191742:lmo7aip_90c';
    var gcse = document.createElement('script');
    gcse.type = 'text/javascript';
    gcse.src = 'https://cse.google.com/cse.js?cx=' + cx;
    var s = document.getElementsByTagName('script')[0];
    setTimeout("document.getElementById('gsc-i-id1').focus();", 1000);
    s.parentNode.insertBefore(gcse, s);
})();

```

My command:

```julia
SashaChernykh@DESKTOP-0G54NVG E:\Киролайна
$ eslint --fix --no-ignore E:\Киролайна\test.js --debug
  eslint:cli Running on files +0ms
  eslint:glob-util Creating list of files to process. +40ms
  eslint:cli-engine Processing E:\Киролайна\test.js +8ms
  eslint:cli-engine Linting E:\Киролайна\test.js +2ms
  eslint:config Constructing config for E:\Киролайна\test.js +0ms
  eslint:config Using .eslintrc and package.json files +4ms
  eslint:config Loading E:\Киролайна\.eslintrc.js +1ms
  eslint:config-file Loading JS config file: E:\Киролайна\.eslintrc.js +1ms
  eslint:config-file Loading C:\Users\SashaChernykh\AppData\Roaming\npm\node_modules\eslint\conf\eslint.json +251ms
  eslint:config-file Loading JSON config file: C:\Users\SashaChernykh\AppData\Roaming\npm\node_modules\eslint\conf\eslint.json +1ms
  eslint:config Using E:\Киролайна\.eslintrc.js +47ms
  eslint:config Merging command line environment settings +1ms
  eslint:config-ops Apply environment settings to config +1ms
  eslint:config-ops Creating config for environment browser +0ms
  eslint:config-ops Creating config for environment es6 +1ms

  eslint:cli-engine Linting code for E:\Киролайна\test.js (pass 1) +1ms
  eslint:cli-engine Generating fixed text for E:\Киролайна\test.js (pass 1) +90ms
  eslint:text-fixer Applying fixes +1ms
  eslint:text-fixer Found fixes to apply +1ms
  eslint:cli-engine Linting code for E:\Киролайна\test.js (pass 2) +0ms
  eslint:cli-engine Generating fixed text for E:\Киролайна\test.js (pass 2) +31ms
  eslint:text-fixer Applying fixes +1ms
  eslint:text-fixer No fixes to apply +0ms
  eslint:cli-engine Linting complete in: 444ms +0ms
  eslint:cli Fix mode enabled - applying fixes +1ms
```

My script after fixing:

```js
(function () {
    var cx = '013024336414733191742:lmo7aip_90c';
    var gcse = document.createElement('script');
    gcse.type = 'text/javascript';
    gcse.src = 'https://cse.google.com/cse.js?cx=' + cx;
    var s = document.getElementsByTagName('script')[0];
    setTimeout('document.getElementById(\'gsc-i-id1\').focus();', 1000);
    s.parentNode.insertBefore(gcse, s);
}());
```

### 6. Environment

**Operating system and version:**
Windows 32-bit 10.0.14393
**node:**
v7.3.0
**npm:**
3.10.9
**eslint:**
v3.12.2

Thanks.

**Comment by platinumazure** [comment]
The rule is working as designed. Using a string in `setTimeout` results in the engine calling `eval`, hence triggering the rule `no-implied-eval`.

This is what you should do instead:

```js
setTimeout(function () {
  document.getElementById('gsc-i-id1').focus();
}, 1000);
```

**Comment by Kristinita** [comment]
@platinumazure , thanks, it's work in Firefox, Opera and Chrome! But ESLint not fixed automatically similar errors?

Thanks.

**Comment by platinumazure** [comment]
ESLint does not support autofix in that rule because it would require parsing the string and somehow transform it into a function. But it's pretty easy to make the change manually.

**Comment by platinumazure** [comment]
By the way, @Kristinita, you can use the ESLint docs to learn if a rule is autofixable or not.

On the main rule listing (http://eslint.org/docs/rules/), you can look for a wrench icon to the left of some of the rules.

In a specific rule's documentation, you can look at the top (under the page header) to see if the rule is fixable or not. (For example, see http://eslint.org/docs/rules/no-extra-boolean-cast.)

Hope this helps!
